### ISSUE TITLE
Out of memory 

### ISSUE THREAD

**Issue by ChaoLi977** [milesial/Pytorch-UNet#40]
Hi, thanks for your code, but I have some problems when runing the code.
My dataset has 19k images. When I use the function "to_cropped_imgs" to yield the list of (img, mask), it seems my computer's memory runs out. I think the generation of images list uses a lot of memory. Is anyone has some ways to prevent load so large a list to the memory, so that I can train the model with a great deal of images?

**Comment by milesial** [comment]
Hi, the input pipeline uses a generator, so images are loaded and processed on the fly. If you have memory problems, I suggest either reducing the image size, or processing less images in parallel (reduce batch size). In my case, I trained the model with a batch size of 1 ! See also the notes on memory in the README

**Comment by ChaoLi977** [comment]
Thanks for your reply. My GPU memory is enough, I use images sized 512*512 and the batch size is 1, which occupies 2GB GPU. 
But my problem is the memory of the computer, I have 32GB memory. The computer crashes down  when producing the list (img,mask). I also think the generator should load and process images on the fly, it shouldn't take too much memory. But the fact is my computer's memory uses out when processing the 4000th images. I am not sure if it is because the number of my training set is too much.

**Comment by ChaoLi977** [comment]
By the way, I think in each epoch, the "train = get_imgs_and_masks(iddataset['train'], dir_img, dir_mask, img_scale)" will output same result. Since the iddataset['train'] doesn't change every epoch. So I think maybe you can put this step out of the "for" loop. 

I think the line 83 in train.py, the index "i" need to be replaced with another variable, as "i" has been used in line65 and line 66.

**Comment by milesial** [comment]
For the `train = get_imgs_and_masks`, it is needed inside the `for` since after one epoch, the generator is exhausted, so if we want to iterate again on a new epoch we need a new (the same) generator.

For the lines 65 & 66 the `i` variable is local to each line, and doesn't affect the `i` of the for loop. Bad design, but it works so I will keep it as it is.

About the memory, that sounds like a memory leak, python garbage collection should do it's job and delete previous images and temporary variables. If you feel like investigating the issue, you could use a memory profiler to see where the memory is kept.

**Comment by ChaoLi977** [comment]
Actually, when I run the train.py, the i variable is changed to an array after the line 65. It indeed be affected, I don't know which python version you use. My python is 2.7 anaconda.

**Comment by milesial** [comment]
This project was created for python 3 so I would use that instead.

**Comment by ChaoLi977** [comment]
OK, I will use python3 to run it again.
